Data Dictionary case property data types

kaapstorm opened the pull request:
## Product Description :loudspeaker: 

This change makes the `CASE_IMPORT_DATA_DICTIONARY_VALIDATION` feature flag generally available for Advanced Plan subscriptions and higher.

* No UI changes to enable or disable the feature are being added. All Advanced Plan subscribers will have case property data types available in the Data Dictionary.

* Those projects that have case properties with data types set to "Date" or "Multiple Choice" will have validation enabled for those case properties when they import cases.



## Technical Summary :memo:

:t-rex: Jira: [SAAS-18497](https://dimagi.atlassian.net/browse/SAAS-18497)

It is expected that the migration in this PR will conflict with changes waiting to be merged. Expect a follow-up commit to resolve conflicts.

:blowfish: Easier to review by commit.



## Feature Flag :white_flag:

`CASE_IMPORT_DATA_DICTIONARY_VALIDATION`



## Safety Assurance :rescue_worker_helmet:

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

Yes

### QA Plan

No

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SAAS-18497]: https://dimagi.atlassian.net/browse/SAAS-18497?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kaapstorm committed:
Add DATA_DICT_TYPES privilege

kaapstorm committed:
Update backend

kaapstorm committed:
Update frontend

kaapstorm committed:
Update tests

kaapstorm committed:
isort

kaapstorm committed:
Fix column widths

mjriley reviewed (approved):
Looks like some good clean up. No blocking changes.

kaapstorm committed:
Merge branch 'master' into nh/ff/dd_types

kaapstorm committed:
uses_data_type_validation

kaapstorm committed:
Drop unnecessary domain names

kaapstorm committed:
Modernize test

kaapstorm committed:
Break out validation methods

kaapstorm committed:
Drop `fields_to_validate()` util

kaapstorm committed:
Add validation for more data types

kaapstorm committed:
Move column

kaapstorm committed:
Merge branch 'master' into nh/ff/dd_types

kaapstorm commented:
Final changes:
* Fixed the migration conflict (again).
* Moved the "Valid Values" column next to the "Data Type" column. See below:

<img width="1289" height="794" alt="image" src="https://github.com/user-attachments/assets/8473b603-c421-459e-b902-4d0693c9eb80" />

kaapstorm committed:
Fix case property selection for validation

kaapstorm committed:
Nit

kaapstorm committed:
Speed up tests

mjriley commented on corehq/apps/data_dictionary/models.py:
```diff
@@ -166,14 +166,32 @@ def delete(self, *args, **kwargs):
+            self.data_type,
+            lambda x: None
+        )
+        return method(value)
```
I love factoring out the old code into the validation methods, but especially since our default validation is a no-op, I'm curious why we opted for a `get` approach over:
```
if self.data_type == self.DataType.DATA.value:
  self._validate_date(value)
elif self.data_type == self.DataType.SELECT.value:
  self._validate_select(value)
```
Are we anticipating many more validation methods in the future?

Also, nit: it doesn't look like the validation methods are meant to return a value, so I'm curious if the `return method(value)` statement should instead forego the `return`.

kaapstorm replied:
:+1:  55977607bff

mjriley commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -322,8 +326,8 @@ def _populate_updated_fields(self, row):
-            if update_field_name in self.fields_to_validate:
-                case_property = self.fields_to_validate[update_field_name]
+            if self.field_to_case_property:
+                case_property = self.field_to_case_property[update_field_name]
```
This doesn't seem to be a direct translation of the previous statement. The previous version was checking that the key was present in `fields_to_validate` before accessing it. The changed version appears to introduce the possibility that accessing `update_field_name` will generate a `KeyError`. Is this something we need to handle?

mjriley commented on corehq/apps/case_importer/exceptions.py:
```diff
@@ -112,6 +112,28 @@ class InvalidDate(CaseRowError):
 
 
+class InvalidNumber(CaseRowError):
+    title = gettext_noop('Invalid Number')
```
I realize this is following the prior `Error` examples, but any idea why `gettext_noop` is being used here rather than `gettext_lazy`?

mjriley reviewed (approved):
Looks good, no blocking comments

kaapstorm committed:
Validation methods don't return a value

kaapstorm committed:
Move function back to util module.

kaapstorm committed:
Use parametrized tests for multiple values

kaapstorm committed:
Improve phone number validation

kaapstorm committed:
Use static methods where possible

kaapstorm committed:
Choice `value` attribute not necessary

kaapstorm commented on corehq/apps/data_dictionary/models.py:
```diff
@@ -166,14 +166,32 @@ def delete(self, *args, **kwargs):
+        )
+        return method(value)
+
+    def _validate_date(self, value):
```
d68f7caf46a

kaapstorm commented on corehq/apps/data_dictionary/tests/test_models.py:
```diff
@@ -74,3 +89,57 @@ def test_check_invalid_choice(self):
+            with pytest.raises(exceptions.InvalidGPS):
+                self.gps_property.check_validity(value)
+
+    def test_check_valid_phone_number(self):
```
Improved phone number validation. 168a3f976b5

kaapstorm committed:
Fix test

kaapstorm committed:
Clearer that validation is only on import

This is already stated in the help text, who ever looks at the help
text? The UI design gives the column heading as "Valid Values (for
import)" but in a reasonably-sized browser window, the line gets broken
at "Valid Values (for // import)" and it looks awkward, so I went with a
shorter column heading.

mjriley commented on corehq/apps/data_dictionary/tests/test_models.py:
```diff
@@ -12,134 +10,114 @@
+    '42.123 -71.456 100',
+    '42.123 -71.456 100 5',
+    '-42.123 -71.456 -10 5.5',  # Negative altitude, decimal precision
+    '42123 -71456',  # Note: Validation does not check degrees
```
Appreciate the comments in some of these parameterized lists. I can definitely see them helping to understand why particular values were chosen

mjriley commented on corehq/apps/data_dictionary/tests/test_models.py:
```diff
@@ -12,134 +10,114 @@
+    # check_validity raises exception on invalid date so just
+    # calling it and not getting an exception indicates validity
+    prop = CaseProperty(data_type=CaseProperty.DataType.DATE)
+    prop.check_validity(valid_date)
```
So, one piece of feedback Biyeun left me was that standalone functions in model or view classes can tend to get lost over time -- it's not clear what functionality they belong to, so they can get shuffled around. I agreed with this feedback, but I wonder if that also should extend to tests on generic modules like models. Testing this module might involve testing for several models, and I could see it being hard to follow when they're all standalone functions like these. But then again, I'm also most comfortable with the class-based test syntax, so just throwing it out there

mjriley commented on corehq/apps/data_dictionary/tests/test_models.py:
```diff
@@ -74,3 +89,57 @@ def test_check_invalid_choice(self):
+        self.number_property.check_validity('42')
+        self.number_property.check_validity('3.14')
+        self.number_property.check_validity('-10')
+        self.number_property.check_validity('-3.14')
```
As long as the test suites are scoped well, I don't tend to care about the length of test modules, so I would have still written out separate cases (and suffered the duplication). That said, parameterized tests seem like a decent middle ground, where they fix the early termination issue.

kaapstorm committed:
Move tests back into a class

kaapstorm committed:
Single quotes
